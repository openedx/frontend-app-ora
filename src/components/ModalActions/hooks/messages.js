import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  finishLater: {
    id: 'ora-mfe.ModalActions.simpleAction.finishLater',
    defaultMessage: 'Save for later',
    description: 'Save for later (close) button text',
  },
  submitResponse: {
    id: 'ora-mfe.ModalActions.submitResponse',
    defaultMessage: 'Submit response',
    description: 'Submit button text',
  },
  submittingResponse: {
    id: 'ora-mfe.ModalActions.submittingResponse',
    defaultMessage: 'Submitting response',
    description: 'Submit button text while submitting',
  },
  responseSubmitted: {
    id: 'ora-mfe.ModalActions.responseSubmitted',
    defaultMessage: 'Response submitted',
    description: 'Submit button text after successful submission',
  },
  savingResponse: {
    id: 'ora-mfe.ModalActions.savingResponse',
    defaultMessage: 'Saving response',
    description: 'Save for later button text while saving',
  },
});

export default messages;
