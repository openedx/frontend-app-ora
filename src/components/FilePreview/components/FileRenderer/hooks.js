import { useState } from 'react';
import { errorStatuses, errorMessages, renderers } from '../constants';
import { getFileType } from '../utils';
import messages from './messages';

/**
 * component hooks
 */
export const useRenderData = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useState(null);
  const [isLoading, setIsLoading] = useState(true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    errorMessages[errorStatus] || errorMessages[errorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = renderers[getFileType(file.fileName)];
  const rendererProps = {
    fileName: file.fileName,
    url: file.fileUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
