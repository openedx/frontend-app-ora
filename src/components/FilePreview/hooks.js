import { useKeyedState, StrictDict } from '@edx/react-unit-test-utils';

import {
  PDFRenderer,
  ImageRenderer,
  TXTRenderer,
} from 'components/FilePreview/BaseRenderers';

import messages from './messages';

export const ErrorStatuses = StrictDict({
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  conflict: 409,
  serverError: 500,
});

export const FileTypes = StrictDict({
  pdf: 'pdf',
  jpg: 'jpg',
  jpeg: 'jpeg',
  png: 'png',
  bmp: 'bmp',
  txt: 'txt',
  gif: 'gif',
  jfif: 'jfif',
  pjpeg: 'pjpeg',
  pjp: 'pjp',
  svg: 'svg',
});

/**
 * Config data
 */
export const RENDERERS = {
  [FileTypes.pdf]: PDFRenderer,
  [FileTypes.jpg]: ImageRenderer,
  [FileTypes.jpeg]: ImageRenderer,
  [FileTypes.bmp]: ImageRenderer,
  [FileTypes.png]: ImageRenderer,
  [FileTypes.txt]: TXTRenderer,
  [FileTypes.gif]: ImageRenderer,
  [FileTypes.jfif]: ImageRenderer,
  [FileTypes.pjpeg]: ImageRenderer,
  [FileTypes.pjp]: ImageRenderer,
  [FileTypes.svg]: ImageRenderer,
};

export const SUPPORTED_TYPES = Object.keys(RENDERERS);

export const ERROR_STATUSES = {
  [ErrorStatuses.notFound]: messages.fileNotFoundError,
  [ErrorStatuses.serverError]: messages.unknownError,
};

/**
 * Util methods and transforms
 */
export const getFileType = (fileName) => fileName.split('.').pop()?.toLowerCase();
export const isSupported = (file) => SUPPORTED_TYPES.includes(
  getFileType(file.name),
);

export const stateKeys = StrictDict({
  errorStatus: 'errorStatus',
  isLoading: 'isLoading',
});

/**
 * component hooks
 */
export const renderHooks = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useKeyedState(stateKeys.errorStatus, null);
  const [isLoading, setIsLoading] = useKeyedState(stateKeys.isLoading, true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    ERROR_STATUSES[errorStatus] || ERROR_STATUSES[ErrorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = RENDERERS[getFileType(file.name)];
  const rendererProps = {
    fileName: file.name,
    url: file.downloadUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
