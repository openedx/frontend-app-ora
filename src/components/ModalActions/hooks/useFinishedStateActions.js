import { useGlobalState, useTrainingStepIsCompleted } from 'hooks/app';
import {
  useHasSubmitted,
  useSubmittedAssessment,
} from 'hooks/assessment';
import { useViewStep } from 'hooks/routing';
import {
  useStartStepAction,
  useLoadNextAction,
  useCloseModalAction,
} from 'hooks/actions';
import { stepNames, stepStates } from 'constants/index';

const useFinishedStateActions = () => {
  const step = useViewStep();
  const globalState = useGlobalState({ step });
  const hasSubmitted = useHasSubmitted();
  const startStepAction = useStartStepAction(step);
  const submittedAssessment = useSubmittedAssessment();
  const loadNextAction = useLoadNextAction();
  const trainingStepIsCompleted = useTrainingStepIsCompleted();

  const stepState = globalState.activeStepState;

  const exitAction = useCloseModalAction();

  if (!hasSubmitted) {
    if (step === stepNames.studentTraining && trainingStepIsCompleted) {
      return { primary: startStepAction, secondary: exitAction };
    }
    return null;
  }

  // assessment finished state
  if (submittedAssessment) {
    const { activeStepName } = globalState;
    if (activeStepName === stepNames.done) {
      return { primary: startStepAction };
    }
    if (activeStepName === stepNames.staff) {
      return { primary: exitAction };
    }
    // finished and moved to next step
    if ([stepNames.submission || stepNames.self].includes(step)) {
      return { primary: startStepAction, secondary: exitAction };
    }
    if (step !== activeStepName) {
      // next step is available
      if (stepState === stepStates.inProgress) {
        return { primary: startStepAction, secondary: exitAction };
      }
      // next step is not available
      return null;
    }

    // finished current assessment but not current step
    if (stepState === stepStates.inProgress) {
      return { primary: loadNextAction, secondary: exitAction };
    }
    // finished current assessment, but not step
    // and there are no more assessments available for the current step
    return { primary: exitAction };
  }
  // submission finished state
  return { primary: startStepAction, secondary: exitAction };
};

export default useFinishedStateActions;
