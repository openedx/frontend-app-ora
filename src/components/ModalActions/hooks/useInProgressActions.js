import { useGlobalState, useStepInfo } from 'hooks/app';
import { useHasSubmitted } from 'hooks/assessment';
import { useViewStep } from 'hooks/routing';
import { stepNames, stepStates } from 'constants/index';

import {
  useCloseModalAction,
  useLoadNextAction,
  useStartStepAction,
  useSubmitResponseAction,
} from 'hooks/actions';

const useInProgressActions = ({ options }) => {
  const step = useViewStep();
  const globalState = useGlobalState({ step });
  const hasSubmitted = useHasSubmitted();
  const startStepAction = useStartStepAction(step);
  const stepInfo = useStepInfo();

  const exitAction = useCloseModalAction();
  const activeSubmissionConfig = {
    primary: useSubmitResponseAction({ options }),
    secondary: exitAction,
  };
  const loadNextAction = useLoadNextAction();

  if (globalState.hasReceivedFinalGrade) {
    return { primary: startStepAction, secondary: exitAction };
  }

  // finished state
  if (hasSubmitted) { return null; }
  if (globalState.activeStepState !== stepStates.inProgress) {
    return null;
  }

  // current step is in-progress and unblocked
  if (globalState.stepState === stepStates.inProgress) {
    // only submission step has inProgress modal actions.
    // assessment step inProgress actions are in the Assessment
    // graded step is never inProgress
    return step === stepNames.submission ? activeSubmissionConfig : null;
  }

  // current step is in progress, but you are not viewing it currently
  const { activeStepName } = globalState;
  // if current step has more assessments
  if (
    [stepNames.peer, stepNames.studentTraining].includes(activeStepName)
    && stepInfo[activeStepName].numberOfAssessmentsCompleted
  ) {
    return { primary: loadNextAction, secondary: exitAction };
  }
  // current step has only one assessment, or is on first one
  return { primary: startStepAction, secondary: exitAction };
};

export default useInProgressActions;
