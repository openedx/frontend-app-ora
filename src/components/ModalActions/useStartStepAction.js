import { useNavigate, useParams } from 'react-router-dom';

import { useIntl } from '@edx/frontend-platform/i18n';

import { stepNames, stepRoutes } from 'data/services/lms/constants';
import { useActiveStepName } from 'data/services/lms/hooks/selectors';
import messages from './messages';

const useStartStepAction = (viewStep) => {
  const { formatMessage } = useIntl();
  const navigate = useNavigate();
  const { courseId, xblockId } = useParams();

  const stepName = useActiveStepName();

  if (viewStep === stepNames.done
    || stepName === stepNames.waiting
    || [stepNames.submission, stepNames.staff].includes(stepName)) {
    return null;
  }

  const onClick = () => navigate(`/${stepRoutes[stepName]}/${courseId}/${xblockId}`);

  const startMessages = {
    [stepNames.studentTraining]: messages.startTraining,
    [stepNames.self]: messages.startSelf,
    [stepNames.peer]: messages.startPeer,
    [stepNames.done]: messages.viewGrades,
  };
  if ( startMessages[stepName] ) {
    return { children: formatMessage(startMessages[stepName]), onClick };
  }
  console.error(`No start message for step ${stepName}`)
  return { children: formatMessage(startMessages[stepNames.done]), onClick }
};

export default useStartStepAction;
