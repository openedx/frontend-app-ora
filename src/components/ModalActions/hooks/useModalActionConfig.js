import { stepNames, stepStates } from 'constants/index';

import { useGlobalState, useStepInfo } from 'hooks/app';
import { useHasSubmitted } from 'hooks/assessment';
import { useViewStep } from 'hooks/routing';
import {
  useCloseModalAction,
  useLoadNextAction,
  useStartStepAction,
} from 'hooks/actions';

import useFinishedStateActions from './useFinishedStateActions';
import useInProgressActions from './useInProgressActions';

const useModalActionConfig = ({ options }) => {
  const step = useViewStep();
  const globalState = useGlobalState({ step });
  const hasSubmitted = useHasSubmitted();
  const finishedStateActions = useFinishedStateActions();
  const inProgressActions = useInProgressActions({ options });
  const stepInfo = useStepInfo();

  const loadNextAction = useLoadNextAction();
  const startStepAction = useStartStepAction();
  const exitAction = useCloseModalAction();

  if (globalState.hasReceivedFinalGrade) {
    return step === stepNames.done
      ? { primary: exitAction }
      : { primary: startStepAction, secondary: exitAction };
  }
  // finished state
  if (hasSubmitted) {
    if (globalState.activeStepState === stepStates.waitingForPeerGrades) {
      return { primary: stepInfo.peer?.isWaitingForSubmissions ? null : loadNextAction, secondary: exitAction };
    }
    if (globalState.activeStepState !== stepStates.inProgress) {
      return { primary: exitAction };
    }
    return finishedStateActions;
  }

  // In Progress states
  if (globalState.activeStepState === stepStates.inProgress) {
    return inProgressActions;
  }

  // Graded step
  if (step === stepNames.done) {
    return { primary: exitAction };
  }
  return null;
};

export default useModalActionConfig;
