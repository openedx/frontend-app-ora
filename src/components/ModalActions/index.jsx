import React from 'react';
import PropTypes from 'prop-types';

import { Skeleton } from '@edx/paragon';

import ActionButton from 'components/ActionButton';
import ConfirmDialog from 'components/ConfirmDialog';

import { useIsPageDataLoading } from 'hooks/app';

import useModalActionConfig from './hooks/useModalActionConfig';

const ModalActions = ({ options }) => {
  const actions = useModalActionConfig({ options });
  const isPageDataLoading = useIsPageDataLoading();
  const { primary, secondary } = actions || {};

  const actionButton = (variant, btnProps) => (
    <ActionButton key={variant} {...{ ...btnProps, variant }} />
  );

  const customWrapper = ({ children }) => (
    <div className="w-100 h-100">
      {children}
    </div>
  );

  if (isPageDataLoading) {
    return (<Skeleton className="mt-2" wrapper={customWrapper} />);
  }

  const renderedActions = [];
  if (secondary) {
    if (secondary.confirmProps) {
      renderedActions.push(actionButton('outline-primary', secondary.action));
      renderedActions.push(<ConfirmDialog key="confirm-primary" {...secondary.confirmProps} />);
    } else {
      renderedActions.push(actionButton('outline-primary', secondary.action));
    }
  }
  if (primary) {
    if (primary.confirmProps) {
      renderedActions.push(actionButton('primary', primary.action));
      renderedActions.push(<ConfirmDialog key="confirm-primary" {...primary.confirmProps} />);
    } else {
      renderedActions.push(actionButton('primary', primary.action));
    }
  }

  return (
    <div className="mt-2">
      {renderedActions}
    </div>
  );
};
ModalActions.defaultProps = {
  options: {},
};
ModalActions.propTypes = {
  options: PropTypes.shape({
    save: PropTypes.func,
    saveStatus: PropTypes.string,
    submit: PropTypes.func,
    submitStatus: PropTypes.string,
  }),
};
export default ModalActions;
