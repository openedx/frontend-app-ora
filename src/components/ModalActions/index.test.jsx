import { shallow } from '@edx/react-unit-test-utils';

import { useIsPageDataLoading } from 'hooks/app';

import useModalActionConfig from './hooks/useModalActionConfig';

import ModalActions from './index';

jest.mock('components/ActionButton', () => 'ActionButton');
jest.mock('components/ConfirmDialog', () => 'ConfirmDialog');
jest.mock('hooks/app', () => ({
  useIsPageDataLoading: jest.fn(),
}));
jest.mock('./hooks/useModalActionConfig', () => jest.fn());

describe('<ModalActions />', () => {
  const props = {
    options: {},
  };
  beforeEach(() => {
    useIsPageDataLoading.mockReturnValue(false);
  });

  it('render skeleton when page data is loading', () => {
    useIsPageDataLoading.mockReturnValueOnce(true);
    const wrapper = shallow(<ModalActions {...props} />);
    expect(wrapper.snapshot).toMatchSnapshot();
    expect(wrapper.instance.findByType('Skeleton')).toHaveLength(1);
  });

  it('render empty when no actions', () => {
    useModalActionConfig.mockReturnValue({});
    const wrapper = shallow(<ModalActions {...props} />);
    expect(wrapper.snapshot).toMatchSnapshot();
    expect(wrapper.instance.findByType('ActionButton')).toHaveLength(0);
    expect(wrapper.instance.findByType('ConfirmDialog')).toHaveLength(0);
  });

  it('can render primary and secondary without confirm', () => {
    useModalActionConfig.mockReturnValue({
      primary: {
        action: {},
      },
      secondary: {
        action: {},
      },
    });
    const wrapper = shallow(<ModalActions {...props} />);
    expect(wrapper.snapshot).toMatchSnapshot();
    expect(wrapper.instance.findByType('ActionButton')).toHaveLength(2);
    expect(wrapper.instance.findByType('ConfirmDialog')).toHaveLength(0);
  });

  it('can render primary and secondary with confirm', () => {
    useModalActionConfig.mockReturnValue({
      primary: {
        action: {},
        confirmProps: {},
      },
      secondary: {
        action: {},
        confirmProps: {},
      },
    });
    const wrapper = shallow(<ModalActions {...props} />);
    expect(wrapper.snapshot).toMatchSnapshot();
    expect(wrapper.instance.findByType('ActionButton')).toHaveLength(2);
    expect(wrapper.instance.findByType('ConfirmDialog')).toHaveLength(2);
  });
});
