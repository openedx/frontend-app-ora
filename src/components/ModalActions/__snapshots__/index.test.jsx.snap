// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalActions /> can render primary and secondary with confirm 1`] = `
<div
  className="mt-2"
>
  <ActionButton
    key="outline-primary"
    variant="outline-primary"
  />
  <ConfirmDialog
    key="confirm-primary"
  />
  <ActionButton
    key="primary"
    variant="primary"
  />
  <ConfirmDialog
    key="confirm-primary"
  />
</div>
`;

exports[`<ModalActions /> can render primary and secondary without confirm 1`] = `
<div
  className="mt-2"
>
  <ActionButton
    key="outline-primary"
    variant="outline-primary"
  />
  <ActionButton
    key="primary"
    variant="primary"
  />
</div>
`;

exports[`<ModalActions /> render empty when no actions 1`] = `
<div
  className="mt-2"
/>
`;

exports[`<ModalActions /> render skeleton when page data is loading 1`] = `
<Skeleton
  className="mt-2"
  wrapper={[Function]}
/>
`;
